### PR TITLE
3086 - Improve mention and user search

### LINKED ISSUE #3086: Improve user mention & search experience
### Context & proposal

As an effort to provide custom mentions lists, custom suggestions and mentions for non-member users we need to improve our logic and add better checks to find mentions after selecting them from the suggestion list.

### Acceptance criteria

To accept this task we need to complete the following items:

- [ ] Fix the populateMentions logic.
- [ ] Store the mentions after selecting them in the list.
- [ ] Filter out mentions when sending the message to the API.

### PR DESCRIPTION
**Pull request by filbabic**
### 🎯 Goal

Closes #3086 

### 🛠 Implementation details

Added state that keeps the currently selected mentioned users in a set, before sending the message. Added checks to see if they're still in the input before sending it.

Also added checks that match the manual input to see if there's anyone from the channel that we didn't add to the mentions.

### 🧪 Testing

**Testing the new pre-fill logic**:
1. Open any channel and start the mention action.
2. Pick any user from the list.
3. Debug this branch and check if the message before sending has populated `mentionedUserIds`.

**Testing the manual input logic**
1. Open any channel. 
2. Manually write the name of a person you want to mention from the channel. Note: The name has to be 100% match (e.g. "@Filip Babić" for Filip, if he's in the channel).
3. Debug if the message receives new mentions, even without picking a user from the suggestion list.

**Testing that non-matched mention input is removed from mentions**:
1. Open any channel.
2. Start the mention command and **select a user** from the suggestion list.
3. Delete 1 or more characters of the selected user autofill.
4. Debug if the message mention is removed given that it doesn't 100% match the name.

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt [hunk 1]
```diff
@@ -30,18 +30,25 @@ internal fun Message.updateUsers(users: Map<String, User>): Message =
 
 /**
  * Fills [Message.mentionedUsersIds] based on [Message.text] and [Channel.members].
+ *
+ * It combines the users found in the input with pre-set [Message.mentionedUsersIds], in case someone
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt:34]: Let's mention it only searches for channel's members

**Review by adasiewiczr — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [ ] ~UI Changes correct (before & after images)~
- [ ] ~Bugs validated (bugfixes)~
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt [hunk 2]
```diff
@@ -101,8 +103,27 @@ public class MessageInputViewModel @JvmOverloads constructor(
...
+        this.selectedMentions += user
+    }
+
+    /**
+     * Sends a regular message to the channel.
+     *
+     * @param messageText The current message text.
+     * @param messageTransformer Transformer that applies custom changes to the message, before being sent.
+     */
     public fun sendMessage(messageText: String, messageTransformer: Message.() -> Unit = { }) {
```
Comment (hustoj2) [inline stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:121]: I believe the `editMessage()` function also needs these changes, as they could have added a mention while editing.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt [hunk 3]
```diff
@@ -30,18 +30,25 @@ internal fun Message.updateUsers(users: Map<String, User>): Message =
...
+ * @param channel The channel whose members we can check for the mention.
  */
 internal fun Message.populateMentions(channel: Channel) {
     if ('@' !in text) {
         return
     }
 
     val text = text.lowercase()
-    mentionedUsersIds = channel.members.mapNotNullTo(mutableListOf()) { member ->
+    val mentions = mentionedUsersIds.toMutableSet() + channel.members.mapNotNullTo(mutableListOf()) { member ->
```
Comment (hustoj2) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt:45]: This change could also have the benefit of allowing library users to call `ChatDomain.sendMessage()/editMessage()` manually from a custom `MessageSendHandler`, populating any `mentionedUserIds` they'd like.

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt (hunk 2 above)
Comment (filbabic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:121]: Good point! Easy to overlook! :]

I'll add the changes there too!

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt (hunk 3 above)
Comment (filbabic) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt:45]: Yeah, this is the main update to allow mentions to be populated externally + with the input, so it allows for external `mentionedUserIds` to be set.

But it has the added benefit that you described!

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt (hunk 1 above)
Comment (filbabic) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/extensions/Message.kt:34]: done!

File: stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt (hunk 2 above)
Comment (filbabic) [reply stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/viewmodel/MessageInputViewModel.kt:121]: @hustoj2 https://github.com/GetStream/stream-chat-android/pull/3088 This PR will be most probably pushed to main for a patch update! :]

The comment was addressed there, so we'll pull the changes once things are cleared up and merged! :]

**Comment by adasiewiczr** [comment]
Replaced by #3088

**Pull request closed without merging**